Add full SH basis

CHrlS98 opened the pull request:
Add bases `descoteaux07_full `and `tournier07_full` as available SH basis in `dipy/reconst/shm.py`. These bases add odd order spherical harmonics to the symmetric basis and can approximate symmetric and asymmetric spherical functions. New functions are `real_full_sh_mrtrix` (tournier07_full), `real_full_sh_basis` (descoteaux07_full) and `sph_harm_full_ind_list` (used to construct index list of orders and degrees of the complete SH basis).

BramshQamar commented on dipy/reconst/shm.py:
```diff
@@ -231,6 +231,60 @@ def real_sph_harm(m, n, theta, phi):
+    -----------
+    sh_order : int
+        The maximum degree or the spherical harmonic basis.
+    theta : float [0, pi]
```
Not sure if it is the right way to write like this `theta : float [0, pi] ` in the docstring. 
I assume it means theta can take value within the range of `0-pi`.

What do you think? @skoudoro

GuillaumeTh replied:
Based on the documentation of `real_sph_harm`. The docstring looks good.

This `theta : float [0, pi]` looks good to me.

skoudoro replied:
I confirm that `[0, pi]` should be with the description below

GuillaumeTh commented on dipy/reconst/shm.py:
```diff
@@ -865,11 +1023,16 @@ def sf_to_sh(sf, sphere, sh_order=4, basis_type=None, smooth=0.0):
-        ``descoteaux07`` for the Descoteaux 2007 [1]_ basis
+        ``tournier07`` for the symmetric Tournier 2007 [2]_ basis,
+        ``descoteaux07`` for the symmetric Descoteaux 2007 [1]_ basis,
+        ``tournier07_full`` for Tournier 2007 [2]_ basis with odd orders, and
```
Be more precise : `basis with odd orders` -> `basis with odd and even orders`

GuillaumeTh commented on dipy/reconst/shm.py:
```diff
@@ -865,11 +1023,16 @@ def sf_to_sh(sf, sphere, sh_order=4, basis_type=None, smooth=0.0):
+        ``tournier07`` for the symmetric Tournier 2007 [2]_ basis,
+        ``descoteaux07`` for the symmetric Descoteaux 2007 [1]_ basis,
+        ``tournier07_full`` for Tournier 2007 [2]_ basis with odd orders, and
+        ``descoteaux07_full`` for Descoteaux 2007 [1]_ basis with odd orders
```
Be more precise : `basis with odd orders` -> `basis with odd and even orders`

GuillaumeTh commented on dipy/reconst/shm.py:
```diff
@@ -915,11 +1078,16 @@ def sh_to_sf(sh, sphere, sh_order, basis_type=None):
-        ``descoteaux07`` for the Descoteaux 2007 [1]_ basis
+        ``tournier07`` for the symmetric Tournier 2007 [2]_ basis,
+        ``descoteaux07`` for the symmetric Descoteaux 2007 [1]_ basis,
+        ``tournier07_full`` for Tournier 2007 [2]_ basis with odd orders, and
```
Be more precise : `basis with odd orders` -> `basis with odd and even orders`

GuillaumeTh commented on dipy/reconst/shm.py:
```diff
@@ -915,11 +1078,16 @@ def sh_to_sf(sh, sphere, sh_order, basis_type=None):
+        ``tournier07`` for the symmetric Tournier 2007 [2]_ basis,
+        ``descoteaux07`` for the symmetric Descoteaux 2007 [1]_ basis,
+        ``tournier07_full`` for Tournier 2007 [2]_ basis with odd orders, and
+        ``descoteaux07_full`` for Descoteaux 2007 [1]_ basis with odd orders
```
Be more precise : `basis with odd orders` -> `basis with odd and even orders`

GuillaumeTh commented on dipy/reconst/shm.py:
```diff
@@ -961,11 +1129,16 @@ def sh_to_sf_matrix(sphere, sh_order, basis_type=None, return_inv=True,
-        ``descoteaux07`` for the Descoteaux 2007 [1]_ basis
+        ``tournier07`` for the symmetric Tournier 2007 [2]_ basis,
+        ``descoteaux07`` for the symmetric Descoteaux 2007 [1]_ basis,
+        ``tournier07_full`` for Tournier 2007 [2]_ basis with odd orders, and
```
Be more precise : `basis with odd orders` -> `basis with odd and even orders`

GuillaumeTh commented on dipy/reconst/shm.py:
```diff
@@ -961,11 +1129,16 @@ def sh_to_sf_matrix(sphere, sh_order, basis_type=None, return_inv=True,
+        ``tournier07`` for the symmetric Tournier 2007 [2]_ basis,
+        ``descoteaux07`` for the symmetric Descoteaux 2007 [1]_ basis,
+        ``tournier07_full`` for Tournier 2007 [2]_ basis with odd orders, and
+        ``descoteaux07_full`` for Descoteaux 2007 [1]_ basis with odd orders
```
Be more precise : `basis with odd orders` -> `basis with odd and even orders`

frheault commented:
@CHrlS98 can you update your branch with upstream/master, I think this may help with the travis error.
@skoudoro Are these error normal, they are not from this PR.

skoudoro commented:
Hi @frheault,

The Viz errors are due to the new VTK 9 (released 1 week ago). We currently fixing them on FURY as you can see here: https://github.com/fury-gl/fury/pull/245. Feel free to give us some feedback it seems we still have some issue on Linux.

The other errors are quite new and I supposed it is due to the new Numpy and the new Scipy releases 2-3 days ago... I need to look at them.

Thanks @CHrlS98  for your update

Garyfallidis commented on dipy/reconst/shm.py:
```diff
@@ -231,6 +231,60 @@ def real_sph_harm(m, n, theta, phi):
 
 
 def real_sym_sh_mrtrix(sh_order, theta, phi):
```
This is the symmetric version. This needed change too? Will changing this affect other areas of the existing code?

CHrlS98 replied:
Hi. Git failed to diff properly, but the real change is the addition of the function `real_full_sh_mrtrix`. The function `real_sym_sh_mrtrix` is the same as before and therefore these changes shouldn't affect other areas of the existing code.

Garyfallidis commented:
@CHrlS98 thank your for this PR. Extending to the full basis was something that we needed for sometime. I was wondering if you can visualize some spherical functions and show the differences. A short up to the point tutorial were you show examples of symmetric and non symmetric spherical functions would be much welcome and would help reviewing this PR.

Garyfallidis commented on dipy/reconst/shm.py:
```diff
@@ -904,7 +1069,7 @@ def sf_to_sh(sf, sphere, sh_order=4, basis_type=None, smooth=0.0):
 
 
-def sh_to_sf(sh, sphere, sh_order, basis_type=None):
+def sh_to_sf(sh, sphere, sh_order=4, basis_type=None):
```
Not sure why you set default to 4 here.  Can you explain?

CHrlS98 replied:
Hi. The function header said the `sh_order`'s default value was 4, but the function's signature didn't fit the description. I could have changed the comment instead of the function signature.

Garyfallidis commented on dipy/reconst/tests/test_shm.py:
```diff
@@ -377,6 +449,24 @@ def test_sf_to_sh():
+    odf2 = sh_to_sf(odf_sh, sphere, 7, "tournier07_full")
+    assert_array_almost_equal(odf, odf2, 2)
+
+    odf_sh = sf_to_sh(odf, sphere, 7, "descoteaux07_full")
```
Are you still using a ``hemi_icosahedron`` here? I think you should use a full sphere as this is a full (non-symmetric) basis.

CHrlS98 replied:
Yes, the test uses a `hemi_icosahedron` for the full basis. I'll replace it by a full sphere.

CHrlS98 commented:
I updated the tests. A short tutorial showing reconstruction of asymmetric signals using a full SH basis is coming soon.

pep8speaks commented:
Hello @CHrlS98, Thank you for updating !





* In the file [`doc/examples/reconst_sh.py`](https://github.com/dipy/dipy/blob/b6b5006f866355b78cfab6136b542521d4e65898/doc/examples/reconst_sh.py):

> [Line 21:1](https://github.com/dipy/dipy/blob/b6b5006f866355b78cfab6136b542521d4e65898/doc/examples/reconst_sh.py#L21): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 29:1](https://github.com/dipy/dipy/blob/b6b5006f866355b78cfab6136b542521d4e65898/doc/examples/reconst_sh.py#L29): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 60:1](https://github.com/dipy/dipy/blob/b6b5006f866355b78cfab6136b542521d4e65898/doc/examples/reconst_sh.py#L60): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 76:1](https://github.com/dipy/dipy/blob/b6b5006f866355b78cfab6136b542521d4e65898/doc/examples/reconst_sh.py#L76): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file
> [Line 77:1](https://github.com/dipy/dipy/blob/b6b5006f866355b78cfab6136b542521d4e65898/doc/examples/reconst_sh.py#L77): [E402](https://duckduckgo.com/?q=pep8%20E402) module level import not at top of file

CHrlS98 commented:
Hi. I added an example showing how spherical harmonics can be used to reconstruct symmetric and asymmetric signals using symmetric and full bases. I also updated the theory documentation on SH basis to include a paragraph on full SH bases.

skoudoro commented:
Thank you @CHrlS98, it is ready to go. Can you just rebase this PR to make Azure and Travis happy?

Any other comments?

skoudoro commented:
Sorry, I do not think you need a rebase finally.  I will just restart the CI's

skoudoro commented:
thank you @CHrlS98, I'am going ahead and merge this PR
